platforms/core-runtime/service-provider/src/main/java/org/gradle/internal/service/ServiceRegistration.java

```diff
@@ -40,7 +41,18 @@ public interface ServiceRegistration {
      * @param serviceType The service to make visible.
      * @param implementationType The implementation type of the service.
      */
-    void add(Class<?> serviceType, Class<?> implementationType);
+    <S, I extends S> void add(Class<S> serviceType, Class<I> implementationType);
+
+    /**
+     * Adds two services to this registry that share the implementation.
+     * <p>
+     * The implementation class should have a single public constructor, and this constructor can take services to be injected as parameters.
+     *
+     * @param serviceType1 The first service to make visible.
+     * @param serviceType2 The second service to make visible.
+     * @param implementationType The implementation type of the service.
+     */
+    <S, I extends S> void add(Class<S> serviceType1, Class<?> serviceType2, Class<I> implementationType);
```

Comment:
I was thinking about this a lot, but decided that for now it should be a good-enough solution.

The main thing I wanted to avoid was inconsistency. If we put the implementation type first, then we should do so for all `add()` overloads. While it would be okay to flip arguments in the `add(serviceClass, implClass)`, the other `add(serviceClass, instance)` overload already has a lot of usages.

The other reason that I thought it would be nicer to keep `add(serviceClass, instance)` ordering is for cases when the implementation is provided as an anonymous object or [a lambda](https://github.com/gradle/gradle/blob/f86092917e29028f4b13aada532d225875db72d9/platforms/core-execution/build-cache-http/src/main/java/org/gradle/caching/http/internal/HttpBuildCacheServiceServices.java#L31). But looking deeper, it seems it is not that common.

In any case, we don't anticipate that having multi-service implementation should be widespread. This new functionality is mainly to keep compatibility with the existing state of things while providing more safety.

Reply:
> I was thinking about this a lot, but decided that for now it should be a good-enough solution.
> In any case, we don't anticipate that having multi-service implementation should be widespread. This new functionality is mainly to keep compatibility with the existing state of things while providing more safety.

I don't insist. There are other ways to do that, if we find the current approach limiting, e.g. Guice-inspired `bind(...).to(...)`.

Reply:
>  I'd suggest changing the order of arguments to impl, service1, service2

I agree with Mike. It's way more natural to have the varying part at the end.